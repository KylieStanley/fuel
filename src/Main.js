import React from 'react'

const Main = () => {
	return (
		<div className="splash">
			<div className="ovrlay"></div>
			<div className="splash-green"></div>
				<div className="main-text">
					<h1 className="fuel">fuel</h1>
					<h3>Recipes and Nutrition to support your training</h3>
				</div>
		</div>	
	)
}

export default Main