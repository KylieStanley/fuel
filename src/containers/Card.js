import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { Link, withRouter } from 'react-router-dom'
import { addFavorite, removeFavorite, selectCard } from '../actions'
import '../styles/main.scss'


export class Card extends Component {
	toggleFavorite = () => {
		const {
			favorites, recipe, removeFavorite, addFavorite
		} = this.props

		const match = favorites.find(favorite => favorite.url === recipe.url)

		if (match) {
			removeFavorite(recipe)
		} else {
			addFavorite(recipe)
		}
	}

	render() {
		const { favorites, recipe, selectCard } = this.props
		const favorite = favorites.find(favorite => favorite.url === recipe.url)
		const favoriteClass = favorite ? <i className="favorite fas fa-heart" /> : <i className="far fa-heart" />

		return (
			<div className="card">
				<div className="image" style={{ backgroundImage: `url(${recipe.image})` }} />
				<div className="recipe-info">
					<Link to={`${this.props.match.url}/modal`} onClick={selectCard.bind(null, recipe)}>
						<h3 className="recipe-name">{recipe.name}</h3>
					</Link>
					<div className="card-bottom">
						<button type="button" className="fave-btn" onClick={this.toggleFavorite}>
							{favoriteClass}
						</button>
						{recipe.dietLabel.map(label => <p>{label}</p>)}
					</div>
				</div>
			</div>
		)
	}
}

export const mapStateToProps = (state) => ({
	favorites: state.favorites
})

export const mapDispatchToProps = (dispatch) => ({
	addFavorite: (recipe) => dispatch(addFavorite(recipe)),
	removeFavorite: (recipe) => dispatch(removeFavorite(recipe)),
	selectCard: (recipe) => dispatch(selectCard(recipe))
})

Card.propTypes = {
	addFavorite: PropTypes.func.isRequired,
	removeFavorite: PropTypes.func.isRequired,
	selectCard: PropTypes.func.isRequired,
	favorites: PropTypes.arrayOf(PropTypes.object).isRequired,
	recipe: PropTypes.shape({
		name: PropTypes.string,
		yield: PropTypes.number,
		calories: PropTypes.number,
		dietLabel: PropTypes.array,
		url: PropTypes.string,
		image: PropTypes.string,
		source: PropTypes.string,
		ingredients: PropTypes.array
	}).isRequired
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Card))
