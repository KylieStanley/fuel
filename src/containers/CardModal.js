import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { addIngredients } from '../actions'
import '../styles/main.scss'

export const CardModal = (props) => {
	const { recipe } = props
	const ingredientList = recipe.ingredients.map(ingredient => <li key={ingredient}>{ ingredient }</li>)

	return (
		<div className="expanded-card">
			<h2 className="expanded-title">{props.recipe.name}</h2>
			<div className="expanded-info">
				<h3>Serves: {recipe.yield}</h3>
				<h3>Calories per Serving: {recipe.calories}</h3>
				<h3>Source: {recipe.source}</h3>
			</div>
			<div className="ingredients">
				<h2 className="ingredients-title">Ingredients</h2>
				<ul>
					{ ingredientList }
				</ul>
			</div>
			<div className="link">
				<button className="add-btn" type="button" onClick={props.addIngredients.bind(null, props.recipe.ingredients)}>Add Ingredients to shopping List +</button>
				<a
					href={recipe.url}
					target="_blank"
					rel="noopener noreferrer"
				>Go To Full Recipe
					<i className="fas fa-angle-double-right" />
				</a>
			</div>
		</div>
	)
}

CardModal.propTypes = {
	recipe: PropTypes.shape({
		name: PropTypes.string,
		yield: PropTypes.number,
		calories: PropTypes.number,
		dietLabel: PropTypes.array,
		url: PropTypes.string,
		image: PropTypes.string,
		source: PropTypes.string,
		ingredients: PropTypes.array
	}).isRequired
}

export const mapDispatchToProps = (dispatch) => ({
	addIngredients: (ingredients) => dispatch(addIngredients(ingredients))
})

export default connect(null, mapDispatchToProps)(CardModal)
