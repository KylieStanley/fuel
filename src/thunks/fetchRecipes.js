import { isLoading, hasErrored, fetchRecipeSuccess } from '../actions'
import { cleanRecipes } from '../helper'

export const fetchRecipes = (url) => {
	return async (dispatch) => {
		try {
			dispatch(isLoading(true))
			const response = await fetch(url)
			if (!response.ok) {
				throw Error(response.statusText)
			}
			dispatch(isLoading(false))
			const result = await response.json()
			const recipes = cleanRecipes(result.hits)

			dispatch(fetchRecipeSuccess(recipes))
		} catch (err) {
			dispatch(hasErrored(err.message))
		}
	}
}
