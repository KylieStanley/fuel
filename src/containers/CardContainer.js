import React from 'react'
import '../styles/main.scss'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import Card from './Card'


export const CardContainer = (props) => {
	const { recipes, favorites, isLoading } = props
	let selectedRecipes = recipes

	if (props.itemType === 'favorites') {
		selectedRecipes = favorites
	}

	const error = !selectedRecipes.length && selectedRecipes === favorites ? <h2>You have no favorites!</h2> : null

	return (
		<div className="card-container">
			{ isLoading ? <img src="../loading.gif" alt="spinning circle" /> : selectedRecipes.map(recipe => <Card key={recipe.url} recipe={recipe} />) }
			{ error }
		</div>
	)
}

export const mapStateToProps = (state) => ({
	isLoading: state.isLoading,
	recipes: state.recipes,
	favorites: state.favorites
})

CardContainer.propTypes = {
	isLoading: PropTypes.bool.isRequired,
	recipes: PropTypes.arrayOf(PropTypes.object).isRequired,
	favorites: PropTypes.arrayOf(PropTypes.object).isRequired
}

export default connect(mapStateToProps)(CardContainer)
