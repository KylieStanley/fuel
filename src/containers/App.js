import React, { Component } from 'react';
import '../styles/main.scss';
import { Route, withRouter } from 'react-router-dom'
import Splash from '../components/Splash'
import List from '../components/List'
import { CardModal } from '../components/CardModal'
import Main from './Main'
import { ModalContainer, ModalRoute } from 'react-router-modal';
import { connect } from 'react-redux'
import 'react-router-modal/css/react-router-modal.css'


export const App = (props) => {
  return (
    <div className="splash">
      <div className="overlay" />
      <div className="splash-green" />
      <Route exact path='/' component={ Splash } />
      <Route path='/main' component={ Main } />
      <Route path="/favorites" render={() => <List itemType="favorites" />} />
      <Route path="/shopping" render={() => <List itemType="shopping" />} />
      <ModalRoute component={ CardModal } path='*/modal' parentPath={`${props.history.goBack}`} props={{ recipe: props.selectedCard }} />
      <ModalContainer />
    </div>
  );
}

export const mapStateToProps = (state) => ({
  selectedCard: state.selectedCard
})

export default withRouter(connect(mapStateToProps)(App));

