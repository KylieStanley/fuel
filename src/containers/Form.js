import React, { Component } from 'react'
import '../styles/main.scss';
import { connect } from 'react-redux'
import { cleanSearchString } from '../helper'
import { apiId, apiKey } from '../apiKey'
import { fetchRecipes } from '../thunks/fetchRecipes'


export class Form extends Component {
	constructor() {
		super()
		this.state = {
			search: '',
			filter: ''
		}
	}

	handleChange = (e) => {
		const { name, value } = e.target
		this.setState({ 
			[name]: value 
		}, () => {
			if (name === 'filter') {
				this.handleSubmit(e)
			}
		})	
	}

	handleSubmit = (e) => {
		const { search, filter } = this.state
		const dietType = `&diet=${filter}` || null
		e.preventDefault()
		if (search) {
			const searchString = cleanSearchString(search)
			const url = `https://api.edamam.com/search?q=${searchString}&app_id=${apiId}&app_key=${apiKey}&from=0&to=30${dietType}`
			this.props.fetchRecipes(url)
		}
	}

	render() {
		return (
		<form onSubmit={this.handleSubmit}>
		<div>
			<input type="text" name="search" placeholder="Search for Recipes" value={this.state.search} onChange={this.handleChange} />
			<button className="search-btn"><i className="fas fa-search"></i></button>
			</div>
			<div class="styled-select white">
			<select name="filter" onChange={this.handleChange}>  
				<option value="">Filter by Diet</option>
				<option value="high-carb">High Carb</option>
				<option value="balanced">Balanced</option>
				<option value="low-carb">Low Carb</option>
				<option value="high-protein">High Protein</option>
				<option value="high-fiber">High Fiber</option>
			</select>
			</div>
		</form>
		)
	}
}

export const mapDispatchToProps = (dispatch) => ({
	fetchRecipes: (url) => dispatch(fetchRecipes(url))
})


export default connect(null, mapDispatchToProps)(Form)