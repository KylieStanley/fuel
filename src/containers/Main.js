import React, { Component } from 'react'
import Form from './Form'
import CardContainer from '../containers/CardContainer'
import { fetchRecipes } from '../thunks/fetchRecipes'
import { connect } from 'react-redux'
import { apiId, apiKey } from '../apiKey'
import Menu from '../components/Menu'
import PropTypes from 'prop-types'


export class Main extends Component {

	componentDidMount() {
		if (this.props.recipes.length === 0) {
			let url = `https://api.edamam.com/search?q=ma&app_id=${apiId}&app_key=${apiKey}&from=0&to=30`
			this.props.fetchRecipes(url)	
		}
	}

	render() {
		return (
			<div className="main-container">
			<div className="header-menu">
				<div className="main-text">
					<h1 className="fuel">fuel</h1>
					<h3>Recipes and Nutrition to support your training</h3>
				</div>
				<Menu />
				</div>
				<Form />
				<CardContainer />
			</div>
		)
	}
}

export const mapStateToProps = (state) => ({
	recipes: state.recipes
})

export const mapDispatchToProps = (dispatch) => ({
  fetchRecipes: (url) => dispatch(fetchRecipes(url))
})

Main.propTypes = {
	recipes: PropTypes.array.isRequired,
	fetchRecipes: PropTypes.func.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(Main)
