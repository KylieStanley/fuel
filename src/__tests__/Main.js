import React from 'react'
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme'
import { Main, mapStateToProps, mapDispatchToProps } from '../containers/Main'
import { fetchRecipes } from '../thunks/fetchRecipes'


jest.mock('../thunks/fetchRecipes')

describe('Main', () => {
	let wrapper
	let mockRecipes

	beforeEach(() => {
		mockRecipes = []
		wrapper = shallow(<Main recipes={ mockRecipes } fetchRecipes={ jest.fn() } />)
	})

	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot()
	})

	it('should call fetchRecipes when mounted if there are no recipes in store', () => {
		expect(wrapper.instance().props.fetchRecipes).toHaveBeenCalled()
	})

	it('should not call fetchRecipes recipes in store from search', () => {
		wrapper = shallow(<Main recipes={{ name: 'Chicken' }} fetchRecipes={ jest.fn() } />)

		expect(wrapper.instance().props.fetchRecipes).not.toHaveBeenCalled()
	})	

	describe('mapStateToProps', () => {
		it('should return an array of recipes', () => {
			const mockState = {
				recipes: { name: 'chicken' },
				favorites: { name: 'salmon'}
			}
			const expected = { recipes: mockState.recipes }
			const result = mapStateToProps(mockState)
			expect(result).toEqual(expected)
		})
	})

	describe('mapDispatchToProps', () => {
		let mockDispatch

		beforeEach(() => {
			mockDispatch = jest.fn()
		})
    
		it('should call dispatch with the correct params', () => {
			const url = 'recipes.com'
			const actionToDispatch = fetchRecipes(url)
			const result = mapDispatchToProps(mockDispatch)
			result.fetchRecipes(url)
			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
		})
	})
})