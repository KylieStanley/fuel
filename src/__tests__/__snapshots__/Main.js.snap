// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Main
    fetchRecipes={
      [MockFunction] {
        "calls": Array [
          Array [
            "https://api.edamam.com/search?q=ma&app_id=16b95589&app_key=298a32be6ad93c49abf9f146b74966e5&from=0&to=30",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    recipes={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="header-menu"
        >
          <div
            className="main-text"
          >
            <h1
              className="fuel"
            >
              fuel
            </h1>
            <h3>
              Recipes and Nutrition to support your training
            </h3>
          </div>
          <Menu />
        </div>,
        <Connect(Form) />,
        <Connect(CardContainer) />,
      ],
      "className": "main-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="main-text"
            >
              <h1
                className="fuel"
              >
                fuel
              </h1>
              <h3>
                Recipes and Nutrition to support your training
              </h3>
            </div>,
            <Menu />,
          ],
          "className": "header-menu",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h1
                  className="fuel"
                >
                  fuel
                </h1>,
                <h3>
                  Recipes and Nutrition to support your training
                </h3>,
              ],
              "className": "main-text",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "fuel",
                  "className": "fuel",
                },
                "ref": null,
                "rendered": "fuel",
                "type": "h1",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Recipes and Nutrition to support your training",
                },
                "ref": null,
                "rendered": "Recipes and Nutrition to support your training",
                "type": "h3",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="header-menu"
          >
            <div
              className="main-text"
            >
              <h1
                className="fuel"
              >
                fuel
              </h1>
              <h3>
                Recipes and Nutrition to support your training
              </h3>
            </div>
            <Menu />
          </div>,
          <Connect(Form) />,
          <Connect(CardContainer) />,
        ],
        "className": "main-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="main-text"
              >
                <h1
                  className="fuel"
                >
                  fuel
                </h1>
                <h3>
                  Recipes and Nutrition to support your training
                </h3>
              </div>,
              <Menu />,
            ],
            "className": "header-menu",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <h1
                    className="fuel"
                  >
                    fuel
                  </h1>,
                  <h3>
                    Recipes and Nutrition to support your training
                  </h3>,
                ],
                "className": "main-text",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "fuel",
                    "className": "fuel",
                  },
                  "ref": null,
                  "rendered": "fuel",
                  "type": "h1",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Recipes and Nutrition to support your training",
                  },
                  "ref": null,
                  "rendered": "Recipes and Nutrition to support your training",
                  "type": "h3",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
